Sources/_MatchingEngine/Regex/Parse/Parse.swift

```diff
@@ -78,22 +88,28 @@ extension Parser {
   ///     Regex        -> '' | Alternation
   ///     Alternation  -> Concatenation ('|' Concatenation)*
   ///
-  mutating func parse() throws -> AST {
+  mutating func parse() throws -> ParseResult {
```

Comment:
@hamishknight how would this impact things like isolated options, which kick off a recursive parse to end-of-group, but whose capture structure wouldn't nest because of  this?

Reply:
Also, why are we complicating parsing for something that can be trivially derived from the AST?

Reply:
I think this should work fine with isolated options as they effectively become implicit non-capturing groups, which shouldn't affect the capture structure AFAICT.

And yeah, IMO it would be nicer if the capture structure computation was done on the AST after parsing. I'm happy for that to be done as a follow-up tho.

Reply:
Well, I initially did perform this on the AST but I didn't want this trivial operation to have to traverse the AST.  It's just a tradeoff between separation of concerns and efficiency.  I don't have a strong opinion either way and am happy to change this (on vacation right now; maybe later tonight).

Reply:
For efficiency, I wonder if it would be possible to emit the byte code directly from the AST, rather than build up a tree of individually allocated and ref-counted indirect nodes.

(Also, I expect all the AST enum stuff to dwarf these costs, unless we get off of enums for ASTs too. Either way, I'd rather spend the complexity budget on run-time perf)

Reply:
> I wonder if it would be possible to emit the byte code directly from the AST

This, too, is how I implemented it initially but it made things less readable. For example, the mapping behavior of quantification needs top-down contextual information for emission. With an enum tree the logic is easy to read, which IMO is especially important as we rapidly evolve the prototype.

Reply:
Reverted to an AST pass. I would prefer not to prematurely optimize this flow until the user-facing prototype is mature.